src/classes/frDonation.cls

```diff
@@ -144,16 +150,75 @@ public class frDonation extends frModel {
         return isSuccess;
     }
     
-    public void createOpportunityMapping(Map<String, Object> request, String contactId) {
-        String mappingType = String.valueOf(request.get('opportunityRoleMapping'));
-        System.debug('Mapping Type: ' + mappingType);
-        if (mappingType != 'disabled' && mappingType != null) {
-            OpportunityContactRole role = new OpportunityContactRole();
-            role.ContactId = contactId;
-            role.OpportunityId = o.Id;
-            role.Role = mappingType;
-            insert role;
+    public void createOpportunityMapping(Map<String, Object> request) {
+        Boolean mappingDisabled = request.containsKey('opportunityContactMappingDisabled') ?
+            Boolean.valueOf(request.get('opportunityContactMappingDisabled')): true;
+        if (!mappingDisabled) {
+            String funraiseId = String.valueOf(request.get('id'));
+            String supporterId = String.valueOf(request.get('donorId'));
+            
+            Set<String> relatedSupporterFunraiseIds = new Set<String>{supporterId};
+            String fundraiserId = request.containsKey('fundraiserId') ? String.valueOf(request.get('fundraiserId')) : null;
+            if(String.isNotBlank(fundraiserId)) {
+                relatedSupporterFunraiseIds.add(fundraiserId);
+            }
+            String teamCaptainId = request.containsKey('teamCaptainId') ? String.valueOf(request.get('teamCaptainId')) : null;
+            if(String.isNotBlank(teamCaptainId)) {
+                relatedSupporterFunraiseIds.add(teamCaptainId);
+            }
+            Map<String, Contact> relatedContacts = new Map<String, Contact>();
+            for(Contact relatedContact : [SELECT Id, fr_Id__c, 
+                                          (SELECT Id, Role, ContactId FROM OpportunityContactRoles WHERE OpportunityId = :getOpportunityId()) 
+                                          from Contact WHERE fr_Id__c IN :relatedSupporterFunraiseIds]) {
+                relatedContacts.put(relatedContact.fr_Id__c, relatedContact);
+            }
+            
+            List<OpportunityContactRole> newContactRoles = new List<OpportunityContactRole>();
+            OpportunityContactRole donorRole = createRole(relatedContacts.get(supporterId), OPP_ROLE_DONOR);
+            if(donorRole != null) newContactRoles.add(donorRole);
```

Comment:
So I had that in here along with your comment below for having an error message for supporterId, but because of the logic in frWSDonationController, we only ever call this method IF the donor was created successfully so in the unit test I wasn't able to create the scenario where we could hit the lines where supporterId is missing, leading the overall class coverage to be lower than we would prefer (and I took that as a message that if it wasn't possible for me to reach that in the test, then they were unnecessary checks)